text/0031-corejam.md

```diff
@@ -0,0 +1,643 @@
+
+```rust
+fn get_work_storage(key: &[u8]) -> Option<Vec<u8>>;
+fn get_work_storage_len(key: &[u8]) -> Option<u32>;
+fn checkpoint() -> Weight;
+fn weight_remaining() -> Weight;
+fn set_work_storage(key: &[u8], value: &[u8]) -> Result<(), ()>;
+fn remove_work_storage(key: &[u8]);
+fn set_validators(validator_keys: &[ValidatorKey]) -> Result<(), ()>;
+fn set_code(code: &[u8]) -> Result<(), ()>;
+fn assign_core(
+    core: CoreIndex,
+    begin: BlockNumber,
+    assignment: Vec<(CoreAssignment, PartsOf57600)>,
+    end_hint: Option<BlockNumber>,
+) -> Result<(), ()>;
+fn transfer(
+    destination: WorkClass,
+    amount: u128,
+    memo: &[u8],
+    weight: Weight,
+) -> Result<Vec<u8>, ()>;
+```
+
+Read-access to the entire Relay-chain state is allowed. No direct write access may be provided since `accumulate` is untrusted code. `set_work_storage` may fail if an insufficient deposit is held under the Work Class's account.
+
+`set_validator`, `set_code` and `assign_core` are all privileged operations and may only be called by pre-authorized Work Classes.
+
+Full access to a child trie specific to the Work Class is provided through the `work_storage` host functions. Since `accumulate` is permissionless and untrusted code, we must ensure that its child trie does not grow to degrade the Relay-chain's overall performance or place untenable requirements on the storage of full-nodes. To this goal, we require an account sovereign to the Work Class to be holding an amount of funds proportional to the overall storage footprint of its Child Trie. `set_work_storage` may return an error should the balance requirement not be met.
+
+Host functions are provided allowing any state changes to be committed at fail-safe checkpoints to provide resilience in case of weight overrun (or even buggy code which panics). The amount of weight remaining may also be queried without setting a checkpoint. `Weight` is expressed in a regular fashion for a solo-chain (i.e. one-dimensional).
+
+Simple transfers of data and balance between Work Classes are possible by the `transfer` function. This is an entirely synchronous function which transfers the execution over to a `destination` Work Class as well as the provided `amount` into their account.
+
+A new VM is set up with code according to the Work Class's `accumulate` code blob, but with a secondary entry point whose prototype is:
+
+```rust
+fn on_transfer(source: WorkClass, amount: u128, memo: Vec<u8>) -> Result<Vec<u8>, ()>;
+```
+
+During this execution, all host functions above may be used except `checkpoint()`. The operation may result in error in which case all changes to state are reverted, including the balance transfer. (Weight is still used.)
+
+Other host functions, including some to access Relay-chain hosted services such as the Balances and Storage Pallet may also be provided commensurate with this executing on-chain.
```

Comment:
Note: so we will never remove balances pallet from relaychain.

Reply:
Balances wouldn't be hosted by the balances pallet since most of its functionality (extrinsics, ED, holds, freezes) is unneeded.